cluster-autoscaler/cloudprovider/azure/azure_agent_pool.go

```diff
@@ -117,9 +131,35 @@ func (as *AgentPool) MaxSize() int {
 	return as.maxSize
 }
 
+func (as *AgentPool) getAllVirtualMachinesWithCache() ([]compute.VirtualMachine, error) {
+	virtualMachinesStatusCache.mutex.Lock()
+	defer virtualMachinesStatusCache.mutex.Unlock()
+
+	if virtualMachinesStatusCache.lastRefresh.Add(vmInstancesRefreshPeriod).After(time.Now()) {
+		return virtualMachinesStatusCache.virtualMachines, nil
+	}
+
+	vms, err := as.GetVirtualMachines()
+	virtualMachinesStatusCache.virtualMachines = vms
+	virtualMachinesStatusCache.lastRefresh = time.Now()
+	if err != nil {
+		if isAzureRequestsThrottled(err) {
+			klog.Warningf("getAllVirtualMachines: throttling with message %v, would return the cached vms", err)
+			return virtualMachinesStatusCache.virtualMachines, nil
+		}
+
+		return []compute.VirtualMachine{}, err
+	}
+
+	return vms, err
+}
+
 // GetVMIndexes gets indexes of all virtual machines belonging to the agent pool.
 func (as *AgentPool) GetVMIndexes() ([]int, map[int]string, error) {
-	instances, err := as.GetVirtualMachines()
+	virtualMachinesStatusCache.mutex.Lock()
+	defer virtualMachinesStatusCache.mutex.Unlock()
```

Comment:
`virtualMachinesStatusCache` has already been locked within `getAllVirtualMachinesWithCache`, hence it shouldn't be locked again